services/hackernews/hackernews-user-karma.service.js

```diff
@@ -0,0 +1,71 @@
+import Joi from 'joi'
+import { metric } from '../text-formatters.js'
+import { BaseJsonService, NotFound } from '../index.js'
+import { anyInteger } from '../validators.js'
+
+const schema = Joi.object({
+  id: Joi.string(),
+  karma: anyInteger,
+})
+  .allow(null)
+  .required()
+
+export default class HackerNewsUserKarma extends BaseJsonService {
+  static category = 'social'
+
+  static route = {
+    base: 'hackernews/user-karma',
+    pattern: ':userid',
+  }
+
+  static examples = [
+    {
+      title: 'HackerNews User Karma',
+      namedParams: { userid: 'drumstick' },
+      staticPreview: this.render({ userid: 'drumstick', karma: 15536 }),
+    },
+  ]
+
+  static defaultBadgeData = {
+    label: 'HackerNews User Karma',
+    namedLogo: 'ycombinator',
+  }
+
+  static render({ karma, userid }) {
+    const color = karma > 0 ? 'brightgreen' : karma === 0 ? 'orange' : 'red'
+    return {
+      label: userid,
+      message: metric(karma),
+      color,
+      style: 'social',
+    }
+  }
+
+  async fetch({ userid }) {
+    return this._requestJson({
+      schema,
+      url: `https://hacker-news.firebaseio.com/v0/user/${userid}.json`,
+      errorMessages: {
+        404: 'user not found',
+      },
+    })
+  }
+
+  transform({ json }) {
+    if (json == null) {
+      throw new NotFound({ prettyMessage: 'user not found' })
+    }
+
+    const { karma, id } = json
+    return { karma, id }
+  }
+
+  async handle({ userid }) {
+    const json = await this.fetch({ userid })
+    const { karma, id } = this.transform({ json })
+    return this.constructor.render({
+      karma,
+      userid: id,
+    })
+  }
```

Comment:
Is there a difference in the `id` value in the response vs the `userid` value we already have via the route parameter?

If not then perhaps consider doing the null check inline with the same early throw/bail, and then calling the render function (with or without the deconstructing).

It's great to see the application of the common `transform` pattern, especially when we have complex logic that's much more easily tested directly via a unit test. However, in this case the canonical not found scenario can be easily exercised already and we aren't really doing anything else but deconstructing and then building a new object.

Just food for thought, don't feel obligated to make any changes

Reply:
Yep, this makes sense. `transform` does not do much work apart from throwing Error. If it was doing some object manipulation, then it would make sense to use it. 

I think we can remove it.